File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        res = [False]
        def sameTree(r1, r2):
            if not r1 and not r2:
                return True

            if not (r1 and r2):
                return False

            return ((r1.val == r2.val) and 
            (sameTree(r1.left, r2.left) and sameTree(r1.right, r2.right)))


        def preOrder(root):
            if not root:
                return

            if sameTree(root, subRoot):
                res[0] = True
            preOrder(root.left)
            preOrder(root.right)

        
        preOrder(root)

        return res[0]

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        target = subRoot->val;
        targetLeft = subRoot->left != nullptr;
        targetRight = subRoot->right != nullptr;
        findSubRoot(root);
        for (TreeNode* psr : potentialSubRoots) {
            if (checkEquality(psr, subRoot)) return true;
        }
        return false;
    }

    void findSubRoot(TreeNode* root) {
        if (root == nullptr) return;
        if (root->val == target &&
           (root->left != nullptr) == targetLeft &&
           (root->right != nullptr) == targetRight
        ) potentialSubRoots.push_back(root);
        findSubRoot(root->left);
        findSubRoot(root->right);
    }

    bool checkEquality(TreeNode* root1, TreeNode* root2) {
        if (root1 == nullptr && root2 == nullptr) return true;
        if (root1 == nullptr || root2 == nullptr) return false;
        bool equal = root1->val == root2->val;
        return (
            equal && 
            checkEquality(root1->left, root2->left) && 
            checkEquality(root1->right, root2->right)
        );
    }

    int target;
    bool targetLeft;
    bool targetRight;

    vector<TreeNode*> potentialSubRoots;
};